File: examples/client-python/mcp_client.py
"""
MCP Enterprise Server — Python 客户端示例

演示 Python 应用如何通过 REST API 调用 MCP Enterprise Server。
无需额外依赖，使用标准库 urllib。

使用方式：
    python mcp_client.py
    或
    MCP_API_KEY=my-key python mcp_client.py
"""

import json
import os
import sys
import urllib.request
import urllib.error

try:
    import ssl
    ctx = ssl.create_default_context()
except ImportError:
    ctx = None


class McpEnterpriseClient:
    """MCP Enterprise Server Python 客户端"""

    def __init__(self, base_url: str = "http://localhost:8081", api_key: str = None):
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key or os.environ.get("MCP_API_KEY", "default-admin-key")
        self.session_id = None

    def _request(self, method: str, path: str, body: dict = None,
                  with_api_key: bool = True) -> dict:
        """通用 HTTP 请求"""
        url = f"{self.base_url}{path}"
        headers = {"Content-Type": "application/json"}
        if with_api_key:
            headers["X-API-Key"] = self.api_key

        data = json.dumps(body).encode("utf-8") if body else None
        req = urllib.request.Request(url, data=data, headers=headers, method=method)

        try:
            with urllib.request.urlopen(req, context=ctx, timeout=30) as resp:
                return json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            return {"error": f"HTTP {e.code}: {e.read().decode()}"}
        except Exception as e:
            return {"error": str(e)}

    def health(self) -> dict:
        """健康检查"""
        return self._request("GET", "/api/mcp/health")

    def connect(self, client_name: str = "python-demo") -> dict:
        """连接 MCP Server"""
        result = self._request("POST", "/api/mcp/connect",
                                body={"clientName": client_name})
        if result.get("success"):
            self.session_id = result.get("sessionId")
        return result

    def disconnect(self) -> dict:
        """断开连接"""
        if not self.session_id:
            return {"error": "No active session"}
        result = self._request("POST", "/api/mcp/disconnect",
                                body={"sessionId": self.session_id})
        if result.get("success"):
            self.session_id = None
        return result

    def list_tools(self) -> dict:
        """列出所有可用工具"""
        return self._request("GET", "/api/mcp/tools")

    def get_tool(self, name: str) -> dict:
        """获取工具详情"""
        return self._request("GET", f"/api/mcp/tools/{name}")

    def invoke_tool(self, name: str, params: dict) -> dict:
        """调用工具"""
        return self._request("POST", f"/api/mcp/tools/{name}/invoke",
                              body=params)

    def stats(self) -> dict:
        """服务统计"""
        return self._request("GET", "/api/mcp/stats")

    def __enter__(self):
        return self

    def __exit__(self, *args):
        if self.session_id:
            self.disconnect()


def main():
    print("🚀 MCP Enterprise Python Client Demo")
    print("=" * 45)

    with McpEnterpriseClient() as client:
        # 1. 健康检查
        print("\n📡 1. 健康检查...")
        health = client.health()
        print(f"   状态: {health.get('status')}")
        print(f"   工具数: {health.get('toolCount')}")
        print(f"   活跃会话: {health.get('activeSessions')}")

        # 2. 连接服务
        print("\n🔗 2. 连接服务...")
        conn = client.connect()
        print(f"   Session ID: {conn.get('sessionId')}")
        print(f"   服务版本: {conn.get('serverVersion')}")

        # 3. 列出工具
        print("\n🔧 3. 可用工具列表...")
        tools = client.list_tools()
        print(f"   总工具数: {tools.get('total', 0)}")
        for tool in tools.get("tools", []):
            print(f"   - {tool.get('name')} ({tool.get('displayName')}) "
                  f"[{tool.get('category')}]")

        # 4. 查看统计
        print("\n📊 4. 服务统计...")
        stats = client.stats()
        print(f"   工具总数: {stats.get('tools', {}).get('total')}")
        print(f"   活跃会话: {stats.get('sessions', {}).get('active')}")

        # 5. 断开连接
        print("\n👋 5. 断开连接...")
        disc = client.disconnect()
        print(f"   断开成功: {disc.get('success')}")

    print("\n✅ Python 示例运行完成!")


if __name__ == "__main__":
    main()

File: examples/client-nodejs/mcp_client.mjs
/**
 * MCP Enterprise Server — Node.js/TypeScript 客户端示例
 *
 * 演示 Node.js 应用如何通过 REST API 调用 MCP Enterprise Server。
 * 支持 API Key 和 OAuth2 Bearer Token 两种认证方式。
 *
 * 使用方式:
 *   npm install
 *   node mcp_client.mjs
 *   # 或
 *   MCP_API_KEY=my-key MCP_AUTH=Bearer node mcp_client.mjs
 */

const BASE_URL = process.env.MCP_URL || 'http://localhost:8081';
const API_KEY = process.env.MCP_API_KEY || 'default-admin-key';
const AUTH_MODE = process.env.MCP_AUTH || 'api-key'; // 'api-key' | 'bearer'

class McpEnterpriseClient {
  constructor(baseUrl, apiKey) {
    this.baseUrl = baseUrl.replace(/\/+$/, '');
    this.apiKey = apiKey;
    this.sessionId = null;
  }

  /**
   * 通用 HTTP 请求
   */
  async _request(method, path, body = null, useAuth = true) {
    const url = `${this.baseUrl}${path}`;
    const headers = { 'Content-Type': 'application/json' };

    if (useAuth) {
      if (AUTH_MODE === 'bearer') {
        headers['Authorization'] = `Bearer ${this.apiKey}`;
      } else {
        headers['X-API-Key'] = this.apiKey;
      }
    }

    const options = { method, headers };
    if (body) {
      options.body = JSON.stringify(body);
    }

    const response = await fetch(url, options);
    const data = await response.json();

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}: ${data.error || JSON.stringify(data)}`);
    }

    return data;
  }

  /** 健康检查 */
  async health() {
    return this._request('GET', '/api/mcp/health', null, false);
  }

  /** 获取认证信息 */
  async authInfo() {
    return this._request('GET', '/api/auth/login', null, false);
  }

  /** 交换 API Key 为会话令牌 */
  async exchangeToken() {
    const result = await this._request('POST', '/api/auth/exchange');
    if (result.success) {
      this.apiKey = result.token;
      console.log(`   ✅ 已交换为会话令牌（有效期 ${result.expiresIn}s）`);
      console.log(`   📌 后续可用: Authorization: Bearer ${result.token.substring(0, 20)}...`);
    }
    return result;
  }

  /** 验证当前令牌 */
  async verifyToken() {
    return this._request('POST', '/api/auth/verify', null, true);
  }

  /** 获取当前用户信息 */
  async me() {
    return this._request('GET', '/api/auth/me');
  }

  /** 连接 MCP Server */
  async connect(clientName = 'nodejs-demo') {
    const result = await this._request('POST', '/api/mcp/connect', { clientName });
    if (result.success) {
      this.sessionId = result.sessionId;
    }
    return result;
  }

  /** 断开连接 */
  async disconnect() {
    if (!this.sessionId) return { error: 'No active session' };
    const result = await this._request('POST', '/api/mcp/disconnect', {
      sessionId: this.sessionId
    });
    if (result.success) this.sessionId = null;
    return result;
  }

  /** 列出所有工具 */
  async listTools() {
    return this._request('GET', '/api/mcp/tools');
  }

  /** 获取工具详情 */
  async getTool(name) {
    return this._request('GET', `/api/mcp/tools/${encodeURIComponent(name)}`);
  }

  /** 调用工具 */
  async invokeTool(name, params = {}) {
    return this._request('POST', `/api/mcp/tools/${encodeURIComponent(name)}/invoke`, params);
  }

  /** 服务统计 */
  async stats() {
    return this._request('GET', '/api/mcp/stats');
  }
}

async function main() {
  console.log('🚀 MCP Enterprise Node.js Client Demo');
  console.log(`   服务器: ${BASE_URL}`);
  console.log(`   认证模式: ${AUTH_MODE}`);
  console.log('='.repeat(50));

  const client = new McpEnterpriseClient(BASE_URL, API_KEY);

  try {
    // 1. 检查认证状态
    console.log('\n🔐 1. 认证信息...');
    const authInfo = await client.authInfo();
    console.log(`   认证模式: ${authInfo.authMode}`);
    console.log(`   支持的方式: ${authInfo.supportedAuthMethods?.map(m => m.type).join(', ')}`);

    // 2. 交换令牌（仅 API Key 模式）
    if (AUTH_MODE === 'api-key') {
      console.log('\n🔄 2. 交换 API Key → 会话令牌...');
      const exchange = await client.exchangeToken();
      console.log(`   成功: ${exchange.success}`);
    }

    // 3. 验证令牌
    console.log('\n✅ 3. 验证令牌...');
    const verify = await client.verifyToken();
    console.log(`   有效: ${verify.valid}`);

    // 4. 健康检查
    console.log('\n📡 4. 健康检查...');
    const health = await client.health();
    console.log(`   状态: ${health.status}`);
    console.log(`   工具数: ${health.toolCount}`);
    console.log(`   活跃会话: ${health.activeSessions}`);

    // 5. 连接服务
    console.log('\n🔗 5. 连接服务...');
    const conn = await client.connect('nodejs-v2');
    console.log(`   Session ID: ${conn.sessionId}`);
    console.log(`   服务版本: ${conn.serverVersion}`);

    // 6. 列出工具
    console.log('\n🔧 6. 可用工具列表...');
    const tools = await client.listTools();
    console.log(`   总工具数: ${tools.total}`);
    for (const tool of (tools.tools || [])) {
      console.log(`   - ${tool.name} [${tool.category || '通用'}]`);
    }

    // 7. 统计信息
    console.log('\n📊 7. 服务统计...');
    const stats = await client.stats();
    console.log(`   活跃会话: ${stats.sessions?.active}`);

    // 8. 断开连接
    console.log('\n👋 8. 断开连接...');
    const disc = await client.disconnect();
    console.log(`   断开成功: ${disc.success}`);

  } catch (err) {
    console.error('\n❌ 错误:', err.message);
    process.exit(1);
  }

  console.log('\n✅ Node.js 示例运行完成!');
}

// TypeScript 类型定义（方便 TypeScript 用户）
/**
 * @typedef {Object} HealthResponse
 * @property {string} status
 * @property {number} toolCount
 * @property {number} activeSessions
 *
 * @typedef {Object} ToolInfo
 * @property {string} name
 * @property {string} displayName
 * @property {string} category
 * @property {string} description
 */

main().catch(console.error);
